pkgs/build-support/rust/build-rust-crate.nix

```diff
@@ -282,14 +282,48 @@ let makeDeps = dependencies:
 
          tr '\n' ' ' < target/link > target/link_
          LINK=$(cat target/link_)
-       fi
+      fi
 
       mkdir -p target/bin
-      echo "${crateBin}" | sed -n 1'p' | tr ',' '\n' | while read BIN; do
-         if [[ ! -z "$BIN" ]]; then
-           build_bin $BIN
-         fi
+      printf "%s\n" "${crateBin}" | head -n1 | tr -s ',' '\n' | while read -r BIN_NAME BIN_PATH; do
+        # filter empty entries / empty "lines"
+        if [[ -z "$BIN_NAME" ]]; then
+             continue
+        fi
+
+        if [[ -z "$BIN_PATH" ]]; then
+          # heuristic to "guess" the correct source file as found in cargo:
+          # https://github.com/rust-lang/cargo/blob/90fc9f620190d5fa3c80b0c8c65a1e1361e6b8ae/src/cargo/util/toml/targets.rs#L308-L325
+
+          # the first two cases are the "new" default IIRC
+          BIN_NAME_=$(echo $BIN_NAME | sed -e 's/_/-/g')
+          FILES="src/bin/$BIN_NAME_.rs src/bin/$BIN_NAME_/main.rs src/bin/main.rs src/main.rs"
```

Comment:
Should this be an array instead?

Reply:
Probably should be an array instead. I'll dig how to do that with bash.